File: apps/mobile/lib/errors/app_error_reporter.dart
enum AppErrorCategory { bootstrap, flutterFramework, platform, zone }

abstract interface class AppErrorReporter {
  void capture(AppErrorCategory category, Object error, StackTrace stackTrace);
}

class NoopAppErrorReporter implements AppErrorReporter {
  const NoopAppErrorReporter();

  @override
  void capture(
    AppErrorCategory category,
    Object error,
    StackTrace stackTrace,
  ) {}
}

File: apps/web/src/app/app-error-reporter.ts
export type AppErrorCategory =
  | "bootstrap"
  | "render"
  | "window-error"
  | "unhandled-rejection"

export type AppErrorEvent = {
  category: AppErrorCategory
  error: unknown
}

export interface AppErrorReporter {
  capture(event: AppErrorEvent): void
}

export const noopAppErrorReporter: AppErrorReporter = {
  capture: () => undefined,
}

export function installGlobalErrorCapture(
  reporter: AppErrorReporter = noopAppErrorReporter,
): () => void {
  const onError = (event: ErrorEvent) => {
    reporter.capture({ category: "window-error", error: event.error })
  }
  const onUnhandledRejection = (event: PromiseRejectionEvent) => {
    reporter.capture({
      category: "unhandled-rejection",
      error: event.reason,
    })
  }

  window.addEventListener("error", onError)
  window.addEventListener("unhandledrejection", onUnhandledRejection)

  return () => {
    window.removeEventListener("error", onError)
    window.removeEventListener("unhandledrejection", onUnhandledRejection)
  }
}
